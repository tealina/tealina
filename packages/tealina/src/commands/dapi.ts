import {
  asyncFlow,
  concat,
  deepFlat,
  filter,
  flat,
  flow,
  isEmpty,
  map,
  notNull,
  pipe,
} from 'fp-lite'
import path from 'node:path'
import { genIndexProp, genTopIndexProp, genWithWrapper } from '../utils/codeGen'
import { Snapshot, completePath, effectFiles } from '../utils/effectFiles'
import { logResults } from '../utils/logResults'
import { calcTypeFileSnapshot } from '../utils/withTypeFile'
import {
  FullContext,
  Seeds,
  collectContext,
  seeds2kindScope,
  validateInput,
} from './capi'

const toKeyMapTrue = flow(
  map((x: string) => [x, true] as const),
  kvs => new Map(kvs),
)

const getTopIndexSnapshot = (
  xs: { kind: string }[],
  topIndexContent: string[],
): Snapshot =>
  pipe(xs, map(flow(v => v.kind, genTopIndexProp)), (imps): Snapshot => {
    const existImpMap = toKeyMapTrue(imps)
    const remains = topIndexContent.filter(v => !existImpMap.has(v))
    return {
      group: 'api',
      action: 'updated',
      filePath: 'index.ts',
      code: genWithWrapper(remains),
    }
  })

interface KindSummary {
  kind: string
  namePathsArray: string[][]
}

const toIndexSnapshot =
  (kindIndexMap: Map<string, string[]>) =>
  (v: KindSummary): (Snapshot & { kind: string }) | null => {
    const imps = v.namePathsArray.map(pathArr => genIndexProp(pathArr))
    const { kind } = v
    const contents = kindIndexMap.get(kind) ?? []
    const remains = contents.filter(v => !imps.includes(v))
    const filePath = path.join(kind, 'index.ts')
    return isEmpty(remains)
      ? { group: 'api', action: 'deleted', filePath, kind }
      : {
          group: 'api',
          action: 'updated',
          filePath,
          code: genWithWrapper(remains),
          kind,
        }
  }

const calcByKinds =
  ({ kindIndexContentMap, topIndexContent }: FullContext) =>
  (kinds: KindSummary[]) => {
    const kindIndexSnapshots = pipe(
      kinds,
      map(toIndexSnapshot(kindIndexContentMap)),
      filter(notNull),
    )
    const deletedKinds = kindIndexSnapshots.filter(v => v.action == 'deleted')
    if (deletedKinds.length < 1) return kindIndexSnapshots
    return [
      getTopIndexSnapshot(deletedKinds, topIndexContent),
      kindIndexSnapshots,
    ]
  }

const getRelativeFilesSnapshots = (ctx: FullContext): Snapshot[] =>
  pipe(
    seeds2kindScope(ctx.seeds),
    calcByKinds(ctx),
    deepFlat(2),
    map(completePath(ctx.commonOption)),
    concat(calcTypeFileSnapshot(ctx)),
  )

const toApiFileSnapshot = (update: Seeds, apiDir: string): Snapshot => ({
  group: 'api',
  action: 'deleted',
  filePath: `${path.join(apiDir, update.kind, ...update.namePaths)}.ts`,
})

const toTestApiSnapshot = (
  update: Seeds,
  { commonOption: { apiDir, testDir } }: FullContext,
): Snapshot => ({
  group: 'test',
  action: 'deleted',
  filePath: `${path.join(
    testDir,
    path.basename(apiDir),
    update.kind,
    ...update.namePaths,
  )}.test.ts`,
})

const mayWithTestFile = (ctx: FullContext) =>
  ctx.commonOption.withTest
    ? (s: Seeds) => [
        toApiFileSnapshot(s, ctx.commonOption.apiDir),
        toTestApiSnapshot(s, ctx),
      ]
    : (s: Seeds) => [toApiFileSnapshot(s, ctx.commonOption.apiDir)]

const calcSnapshots = (ctx: FullContext): Snapshot[] =>
  pipe(
    getRelativeFilesSnapshots(ctx),
    concat(pipe(ctx.seeds, map(mayWithTestFile(ctx)), flat)),
  )

const deleteApis = asyncFlow(
  validateInput,
  collectContext,
  calcSnapshots,
  effectFiles,
  logResults,
)

export { calcSnapshots, deleteApis }
