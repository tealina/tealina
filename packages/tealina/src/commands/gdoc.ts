import chalk from 'chalk'
import path from 'path'
import { parseDeclarationFile } from '../utils/parseDeclarationFile'
import { TealinaComonOption } from '../utils/options'
import { ensureWrite, loadConfig } from '../utils/tool'
import { getApiTypeFilePath } from '../utils/withTypeFile'
import { isEmpty } from 'fp-lite'

interface GdocOptions extends TealinaComonOption {
  outputDir: string
  tsconfig: string
}

export const startGenerateDoc = async (options: GdocOptions) => {
  const config = await loadConfig(path.resolve(options.configPath))
  const { outputDir, apiDir, tsconfigPath } = options
  const entries = [getApiTypeFilePath({ apiDir, typesDir: config.typesDir })]
  const result = parseDeclarationFile({
    entries,
    tsconfigPath,
  })
  const allRecords = Object.values(result).map(v => Object.values(v))
  if (allRecords.every(isEmpty)) {
    throw new Error(
      'Generate document fail! Make sure you type file is correct',
    )
  }
  const filename = path.join(outputDir, path.basename(apiDir)) + '.json'
  ensureWrite(filename, JSON.stringify(result, null, 2))
  console.log(
    chalk.green(
      'Generate document success!, result save at ' + path.join(filename),
    ),
  )
}
