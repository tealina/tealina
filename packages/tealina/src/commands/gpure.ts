import chalk from 'chalk'
import consola from 'consola'
import {
  asyncPipe,
  concat,
  filter,
  flat,
  flow,
  groupBy,
  isEmpty,
  map,
  notNull,
  pipe,
} from 'fp-lite'
import { writeFile } from 'fs/promises'
import { TealinaComonOption } from '../utils/options'
import { parseSchame } from '../utils/parsePrisma'
import { loadConfig } from '../utils/tool'
import {
  BlockAST,
  MatchForExcludeProp,
  MatchForOptionalChcek as MatchForOptionalCheck,
  MatchForTypeTransform,
  MatheLocate,
  MutationKind,
  PropAST,
} from '../gpure.type'
import { PurifyConfig } from '..'

/**
 * static declaration, inject on demand,
 * for avoid unnecesary reference to Prisma.
 */
const JSON_VALUE_TYPE = [
  '/**',
  '* From https://github.com/sindresorhus/type-fest/',
  '* Matches a JSON object.',
  '* This type can be useful to enforce some input to be JSON-compatible or as a super-type to be extended from. ',
  '*/',
  'export type JsonObject = {[Key in string]?: JsonValue}',
  '',
  '/**',
  '* From https://github.com/sindresorhus/type-fest/',
  '* Matches a JSON array.',
  '*/',
  'export interface JsonArray extends Array<JsonValue> {}',
  '',
  '/**',
  '* From https://github.com/sindresorhus/type-fest/',
  '* Matches any valid JSON value.',
  '*/',
  'export type JsonValue = string | number | boolean | JsonObject ',
]

const ENUMS_BEGIN = [
  '/**',
  '* Enums',
  '*/',
  '// Based on',
  '// https://github.com/microsoft/TypeScript/issues/3192#issuecomment-261720275',
  '',
]
interface PurifyOption {
  input: string
  output: string
  namespace: string
}

const formatComment = (lines: string[]) =>
  isEmpty(lines)
    ? []
    : lines.length == 1
    ? [`/** ${lines[0]} */`]
    : ['/**', ...lines.map(v => ` * ${v}`), ' */']

const isMatch = (a: string, b: string) => a == '*' || a == b

const byMatch = (block: BlockAST, kind: MutationKind) => (v: MatheLocate) =>
  isMatch(v.blockName, block.name) &&
  isMatch(v.keyword, block.keyword) &&
  isMatch(v.kind, kind)

const getSpace = (num: number) => Array(num).fill(' ').join('')

const TabSpace = getSpace(2)

/** https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#model-field-scalar-types */
const justMap = new Map<string, string>([
  ['BigInt', 'bigint'],
  ['Int', 'number'],
  ['Float', 'number'],
  ['Decimal', 'number'],
  ['String', 'string'],
  ['DateTime', 'Date | string'],
  ['Boolean', 'boolean'],
  ['Json', 'JsonValue'],
  ['Bytes', 'Buffer'],
])

type GetTsType = (x: PropAST) => string | null | undefined

const TypeTransformStrategies: GetTsType[] = [
  x => justMap.get(x.type),
  x => (x.type.startsWith('Unsupported') ? 'unknown' : null),
]

const PropsExcludeStategies: ((x: PropAST) => boolean)[] = [
  x => x.attribute.has('ignore'),
  x => x.attribute.has('relation'),
  x => x.kind == 'model',
]

const getFirstMatch = (strategies: GetTsType[]) => (x: PropAST) => {
  for (let i = 0, len = strategies.length; i < len; i++) {
    const v = strategies[i](x)
    if (v) return v
  }
  return 'any'
}

const useWhen = (s: string, isValid: boolean) => (isValid ? s : '')

const defaultConfig: Record<
  MutationKind,
  {
    checkOptional: (prop: PropAST) => boolean
  }
> = {
  CreateInput: {
    checkOptional: prop =>
      prop.modifier == '?' ||
      prop.attribute.has('updatedAt') ||
      prop.attribute.has('default'),
  },
  UpdateInput: {
    checkOptional: prop => true,
  },
}

const propDefaultAttr2comment = (prop: PropAST) => {
  const attr = prop.attribute.get('default')
  if (attr == null) return []
  return [`@default {${attr.slice(1, -1)}}`] // without brackets
}

const prop2ts =
  (option: {
    optinalChchek: (prop: PropAST) => boolean
    transformType: (prop: PropAST) => string
  }) =>
  (prop: PropAST) =>
    [
      ...formatComment(
        prop.comment.public.concat(propDefaultAttr2comment(prop)),
      ),
      [
        prop.name,
        useWhen('?', option.optinalChchek(prop)),
        ':',
        ' ',
        option.transformType(prop),
        useWhen('[]', prop.modifier == '[]'),
      ].join(''),
    ]

const block2ts =
  (option: {
    makeName: (block: BlockAST) => string
    checkIsOptional: (block: BlockAST) => (prop: PropAST) => boolean
    transformType: (block: BlockAST) => (prop: PropAST) => string
    notExclude: (block: BlockAST) => (prop: PropAST) => boolean
  }) =>
  (block: BlockAST) => {
    const commentLines = formatComment(block.comment.public)
    const headLine = [`interface ${option.makeName(block)}{`]
    const propLines = pipe(
      block.props,
      filter(option.notExclude(block)),
      map(
        prop2ts({
          optinalChchek: option.checkIsOptional(block),
          transformType: option.transformType(block),
        }),
      ),
      flat,
      map(v => [TabSpace, v].join('')), //indent
    )
    return [commentLines, headLine, propLines, '}', ''].flat()
  }

const findIsOptionalCheck = (
  block: BlockAST,
  kind: MutationKind,
  matches?: MatchForOptionalCheck[],
) => {
  if (matches == null) return null
  const target = matches.find(byMatch(block, kind))
  return target?.predicate
}

const findBlockSpecificTransform = (
  block: BlockAST,
  kind: MutationKind,
  matches?: MatchForTypeTransform[],
) => {
  if (matches == null) return null
  return matches.find(byMatch(block, kind))?.transform
}

const findBlockSpecificExclude = (
  block: BlockAST,
  kind: MutationKind,
  matches?: MatchForExcludeProp[],
) => {
  if (matches == null) return null
  return matches.find(byMatch(block, kind))?.predicate
}

const toFilterFn =
  (predicates: ((x: PropAST) => boolean)[]) => (prop: PropAST) =>
    !predicates.some(fn => fn(prop))

const toDetermineFn =
  (predicates: ((x: PropAST) => boolean)[]) => (prop: PropAST) =>
    predicates.some(fn => fn(prop))

const makeCopositeTypeBy = (kind: MutationKind) => (prop: PropAST) =>
  prop.kind == 'compositeType' ? [prop.type, kind].join('') : null

const makeMutationTsInterface = (
  kind: MutationKind,
  { overwrite, typeRemap }: PurifyConfig,
) =>
  block2ts({
    makeName: block => [block.name, kind].join(''),
    checkIsOptional: block =>
      toDetermineFn(
        [
          findIsOptionalCheck(block, kind, overwrite?.isOptional),
          defaultConfig[kind].checkOptional,
        ].filter(notNull),
      ),
    transformType: block =>
      getFirstMatch(
        [
          findBlockSpecificTransform(block, kind, overwrite?.transofrmType),
          typeRemap ? (prop: PropAST) => typeRemap(prop.type) : null,
          makeCopositeTypeBy(kind),
          ...TypeTransformStrategies,
        ].filter(notNull),
      ),
    notExclude: block =>
      toFilterFn(
        [
          findBlockSpecificExclude(block, kind, overwrite?.excludeProps),
          ...PropsExcludeStategies,
        ].filter(notNull),
      ),
  })

const withoutBreckets = (x?: string) => (x == null ? null : x.slice(1, -1))

const enum2ts = ({ name, comment, props }: BlockAST): string[] => [
  ...formatComment(comment.public),
  `export const ${name}: {`,
  ...props.map(v =>
    [
      `${TabSpace}${v.name}`,
      withoutBreckets(v.attribute.get('map')) ?? `"${v.name}"`,
    ].join(': '),
  ),
  '}',
  `export type ${name} = (typeof ${name})[keyof typeof ${name}]`,
  '',
]

const ConcernedKeywords = ['model', 'enum', 'type']

const wrapperWith =
  ({ input, namespace }: PurifyOption) =>
  (lines: string[]) =>
    [
      ...formatComment([`Purified mutation types from [schema](${input})`]),
      `export namespace ${namespace} {`,
      '',
      ...lines.map(v => `${TabSpace}${v}`),
      '}',
    ]

const makeTypeCodes = (config: PurifyConfig) => (blockList: BlockAST[]) => {
  const hasJsonValue = blockList.some(v =>
    v.props.some(p => p.type == 'JsonValue'),
  )
  return pipe(
    blockList,
    groupBy(block => block.keyword),
    g =>
      pipe(
        g.get('model') ?? [],
        concat(g.get('type') ?? []),
        map(block =>
          pipe(
            makeMutationTsInterface('CreateInput', config)(block),
            concat(makeMutationTsInterface('UpdateInput', config)(block)),
          ),
        ),
        g.has('enum')
          ? flow(concat([ENUMS_BEGIN]), concat(g.get('enum')!.map(enum2ts)))
          : x => x,
      ),
    flat,
    hasJsonValue ? concat(JSON_VALUE_TYPE) : x => x,
  )
}

const workflow = (input: string, config: PurifyConfig) =>
  asyncPipe(
    parseSchame(input),
    filter(block => ConcernedKeywords.includes(block.keyword)),
    filter(block => !block.attribute.has('ignore')),
    makeTypeCodes(config),
  )

const generatePureTypes = async (option: TealinaComonOption & PurifyOption) => {
  const config = await loadConfig(option)
  const purifyConfig = config.gpure ?? {}
  const lines = await workflow(option.input, purifyConfig)
  return pipe(
    lines,
    wrapperWith(option),
    xs => writeFile(option.output, xs.join('\n')),
    () => chalk.green('Types Generated, save at: ' + option.output),
    consola.success,
  )
}
export { generatePureTypes, workflow }
export type { PurifyConfig }
