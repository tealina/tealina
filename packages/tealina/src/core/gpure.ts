import { filter, flow, map, notNull, unique } from 'fp-lite'
import chalk from 'chalk'
import fs from 'node:fs/promises'
import path from 'node:path'
import {
  BlockRange,
  LineInfo,
  clearSymbol,
  getBlockComment,
  makeLineParser,
  parseSchema,
  toFindable,
} from '../utils/parsePrisma.js'

const ArgsValuePattern = /".*"/
/**
 * static declaration, inject on demand,
 * for avoid unnecesary reference to Prisma.
 */
const JSON_VALUE_TYPE = [
  '\t/**',
  '\t* From https://github.com/sindresorhus/type-fest/',
  '\t* Matches a JSON object.',
  '\t* This type can be useful to enforce some input to be JSON-compatible or as a super-type to be extended from. ',
  '\t*/',
  '\texport type JsonObject = {[Key in string]?: JsonValue}',
  '\t',
  '\t/**',
  '\t* From https://github.com/sindresorhus/type-fest/',
  '\t* Matches a JSON array.',
  '\t*/',
  '\texport interface JsonArray extends Array<JsonValue> {}',
  '\t',
  '\t/**',
  '\t* From https://github.com/sindresorhus/type-fest/',
  '\t* Matches any valid JSON value.',
  '\t*/',
  '\texport type JsonValue = string | number | boolean | JsonObject ',
]

const ENUMS_BEGIN = [
  '\t/**',
  '\t * Enums',
  '\t */',
  '\t// Based on',
  '\t// https://github.com/microsoft/TypeScript/issues/3192#issuecomment-261720275',
  '',
]

/** https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#model-field-scalar-types */
const toTsType = (dslType: string) => {
  switch (dslType) {
    case 'BigInt':
      return 'bigint'
    case 'Int':
    case 'Float':
    case 'Decimal':
      return 'number'
    case 'String':
      return 'string'
    case 'DateTime':
      return 'Date | string'
    case 'Boolean':
      return 'boolean'
    case 'Json':
      return 'JsonValue'
    case 'Bytes':
      return 'Buffer'
    default: //Unsupported('*')
      return 'unknow'
  }
}

interface TypeColumn {
  name: string
  modifier: {
    questionMarke?: boolean
    list?: boolean
  }
  contextual: {
    isModel?: boolean
    isCompositeType?: boolean
    isEnum?: boolean
  }
}

/**
 * in line, single @ attributes
 */
interface AttributesColumn {
  default?: string
  id?: boolean
  objectId?: boolean
  updatedAt?: boolean
  ignore?: boolean
  // relation?: { fields: string[] }
}

interface Property {
  name: string
  type: TypeColumn
  comments: string[]
  attributes: AttributesColumn
}

interface Detail {
  comments: string[]
  name: string
  props: Property[]
  // attributes: RowAttributes
}

interface EnumMemeber {
  name: string
  value: string
  comment: string[]
}

interface EnumDetail {
  comments: string[]
  name: string
  props: EnumMemeber[]
  // attributes: RowAttributes
}

const makeTypeColParser = (
  str: string,
  getContextual: (name: string) => TypeColumn['contextual'],
): TypeColumn => {
  const name = clearSymbol(str)
  return {
    name,
    modifier: {
      questionMarke: str.includes('?'),
      list: str.includes('['),
    },
    contextual: getContextual(name),
  }
}

type InputTypeKind = 'Create' | 'Update'

const makeTypeName = (kind: InputTypeKind, target: string) =>
  [target, kind, 'Input'].join('')

const parseAttrCol = (attrPart: string): AttributesColumn => {
  const attrs = attrPart
    .split('@')
    .map(v => v.trim())
    .filter(v => v.length > 0)
  const DefoIndentifier = 'default('
  const defo = attrs.find(v => v.startsWith(DefoIndentifier))
  return {
    // ...extraRelationInfo(attrs),
    default: defo ? defo.slice(DefoIndentifier.length, -1) : void 0,
    objectId: attrs.includes('db.ObjectId'),
    id: attrs.includes('id'),
    ignore: attrs.includes('ignore'),
    updatedAt: attrs.includes('updatedAt'),
  }
}

const checkIsOptional = (x: Property): boolean => {
  const {
    type: { modifier },
    attributes,
  } = x
  return [
    modifier.questionMarke,
    modifier.list,
    attributes.default,
    attributes.updatedAt,
  ].some(v => !!v) //false,null,undefined
}

const parseModelAndType = (schema: string[]) => {
  const { findBlocks, takeBlockName } = toFindable(schema)
  const typeBlocks = findBlocks('type')
  const modelBlocks = findBlocks('model')
  const enumBlocks = findBlocks('enum')
  const enumNames = enumBlocks.map(takeBlockName)
  const subDocNames = typeBlocks.map(takeBlockName)
  const modelNames = modelBlocks.map(takeBlockName)
  const getContextual = (name: string): TypeColumn['contextual'] => ({
    isEnum: enumNames.includes(name),
    isCompositeType: subDocNames.includes(name),
    isModel: modelNames.includes(name),
  })
  const parseCol = (v: LineInfo): Property => {
    const [attrPart, tailComment] = v.restCol.join(' ').split('///')
    return {
      name: v.fisrtCol,
      type: makeTypeColParser(v.secondCol, getContextual),
      attributes: parseAttrCol(attrPart),
      comments: [...v.lineComment, tailComment].filter(notNull),
    }
  }
  const withoutIgnored = ({ lines }: BlockInfo) =>
    !lines.some(line => line.trimStart().startsWith('@@ignore'))
  const toBlockInfo = (blockRange: BlockRange) => ({
    block: blockRange,
    lines: schema.slice(blockRange.begin + 1, blockRange.end - 1),
  })
  interface BlockInfo {
    block: BlockRange
    lines: string[]
  }
  const parseLine = flow(
    map(makeLineParser()),
    filter(notNull),
    map(parseCol),
    filter(v => !v.attributes.ignore),
  )
  const toDetail = ({ block, lines }: BlockInfo): Detail => ({
    comments: getBlockComment(schema, block),
    name: takeBlockName(block)!,
    props: parseLine(lines),
  })
  const parseEnumLine = flow(
    map(makeLineParser()),
    filter(notNull),
    map(parseEnumCol),
    filter(notNull),
  )
  const toEnumDetail = ({ block, lines }: BlockInfo): EnumDetail => ({
    comments: getBlockComment(schema, block),
    name: takeBlockName(block)!,
    props: parseEnumLine(lines),
  })
  const workFlow = flow(map(toBlockInfo), filter(withoutIgnored), map(toDetail))
  const enumWorkFlow = flow(
    map(toBlockInfo),
    filter(withoutIgnored),
    map(toEnumDetail),
  )
  const models = workFlow(modelBlocks)
  const types = workFlow(typeBlocks)
  const enums = enumWorkFlow(enumBlocks)
  return { models, types, enums }
}

const enumTemplate = (entity: EnumDetail) => [
  formatComment('\t', entity.comments),
  `\texport const ${entity.name}: {`,
  ...entity.props.map(v => [`\t\t${v.name}`, v.value].join(': ')),
  '\t}',
  `\texport type ${entity.name} = (typeof ${entity.name})[keyof typeof ${entity.name}]`,
  '',
]

const getPropComment = ({ comments, attributes }: Property) => [
  ...comments,
  ...(attributes.default ? [`@default {${attributes.default}}`] : []),
]

const formatComment = (indent: string, comments: string[]) => {
  const commentLength = comments.length
  const formatedComment =
    commentLength == 1
      ? [`${indent}/**`, comments[0].trim(), '*/'].join(' ')
      : commentLength > 1
      ? [
          [`${indent}/**`, ...comments.map(v => v.trimStart())].join(
            `\n${indent} * `,
          ),
          `${indent} */`,
        ].join('\n')
      : ''
  return formatedComment
}

const valueFromMapAttribute = (attrs: string[]) => {
  const mapFn = attrs.find(v => v.startsWith('map'))
  if (mapFn == null) return null
  const vs = mapFn.match(ArgsValuePattern)
  return vs != null ? vs[0] : null
}

const parseEnumCol = (v: LineInfo): EnumMemeber | null => {
  const [attrPart, tailComment] = (v.secondCol ?? '').split('///')
  const attrs = attrPart.split('@')
  if (attrs.some(v => v.startsWith('ingnore'))) return null
  return {
    name: v.fisrtCol,
    value: valueFromMapAttribute(attrs) ?? `"${v.fisrtCol}"`,
    comment: [...v.lineComment, tailComment].filter(notNull),
  }
}

const makeTsTransformer =
  (calcIsOptional: (p: Property) => boolean) =>
  (kind: InputTypeKind) =>
  (p: Property) => {
    const { type } = p
    const { modifier, contextual } = type
    const tsType = contextual.isCompositeType
      ? makeTypeName(kind, type.name)
      : contextual.isEnum
      ? type.name
      : toTsType(type.name)
    const result = [
      '\t\t',
      p.name,
      calcIsOptional(p) ? '?' : '',
      ': ',
      contextual.isEnum ? `${p.type.name}` : tsType,
      modifier.list ? '[]' : '',
    ].join('')
    const formatedComment = formatComment('\t\t', getPropComment(p))
    return formatedComment.length > 1
      ? [formatedComment, result].join('\n')
      : result
  }

const withoutId = (p: Property): boolean =>
  !(p.attributes.id && p.attributes.default != null)

const makeMutationTsInterface =
  (config: PurifyOption, kind: InputTypeKind) => (detail: Detail) => {
    const { name, props, comments } = detail
    const transmform2ts = makeTsTransformer(
      kind == 'Update' ? p => true : checkIsOptional,
    )
    const pureProps = props
      .filter(withoutId)
      .filter(v => !v.type.contextual.isModel)
    const lines = pureProps.map(transmform2ts(kind))
    return [
      ...(comments.length ? [formatComment('\t', comments)] : []),
      `\texport interface ${makeTypeName(kind, name)} {`,
      ...lines,
      '\t}',
      '',
    ]
  }

type ParsedType = ReturnType<typeof parseModelAndType>

const makePureTypes = (
  { models, types, enums }: ParsedType,
  config: PurifyOption,
) => {
  const makeCreateInputInterface = makeMutationTsInterface(config, 'Create')
  const makeUpdateInputInterface = makeMutationTsInterface(config, 'Update')
  const makeInterfaces = (m: Detail): string[] => [
    ...makeCreateInputInterface(m),
    ...makeUpdateInputInterface(m),
  ]
  const pureInterface = [
    ...models.map(makeInterfaces),
    ...types.map(makeInterfaces),
    ...(enums.length > 0 ? ENUMS_BEGIN : []),
    ...enums.map(enumTemplate),
  ]
    .flat()
    .join('\n')
  return pureInterface
}

const makeFullContent = (
  { namespace, output, input }: PurifyOption,
  pureInterface: string,
  hasJsonType: boolean,
) => {
  const cwd = process.cwd()
  const relative = path.relative(path.dirname(output), cwd)
  const fullInput = path.resolve(input)
  const fromPath = path.join(relative, fullInput.slice(cwd.length))
  const comment = [
    '/**',
    ` * Purified prisma mutation types from [schema](${fromPath})\\`,
    ' * Generated by command ```tealina gpure```\\',
    ' */',
  ]
  const content = [
    ...comment,
    `export namespace ${namespace} {`,
    pureInterface,
    ...(hasJsonType ? [JSON_VALUE_TYPE.join('\n')] : []),
    '}',
  ].join('\n')
  return content
}

interface PurifyOption {
  input: string
  output: string
  namespace: string
}

const checkHasJsonType = ({ models, types }: ParsedType) => {
  const findFlow = flow(
    (v: Detail) => v.props,
    map(p => p.type.name),
    unique,
    xs => xs.some(v => v == 'Json'),
  )
  const hasJsonType = models.some(findFlow) || types.some(findFlow)
  return hasJsonType
}

const generatePureTypes = async (config: PurifyOption) => {
  const schema = await parseSchema(config.input)
  const details = parseModelAndType(schema)
  const content = makePureTypes(details, config)
  const fullContent = makeFullContent(
    config,
    content,
    checkHasJsonType(details),
  )
  await fs.writeFile(config.output, fullContent)
  console.log(chalk.green(`Generated success! output path : ${config.output}`))
}

export {
  generatePureTypes,
  parseModelAndType,
  makePureTypes,
  JSON_VALUE_TYPE,
  ENUMS_BEGIN,
}
