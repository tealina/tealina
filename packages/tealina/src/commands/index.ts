import { cac } from 'cac'
import { createApis } from './capi'
import { deleteApis } from './dapi'
import { generatePureTypes } from './gpure'
import { syncApiByFile } from './sapi'
import { registerGlobalOption } from '../utils/options'
import { startGenerateDoc } from './gdoc'

const cli = cac('tealina')

registerGlobalOption(cli)

cli
  .command('capi [route] [alias]', 'Create API')
  .option('-t,--template-alias <string>', `[string] template alias`)
  .option('--with-test', '[boolean] Generate test file too')
  .option(
    '--by-model',
    `[boolean] create batch APIs by model name in schema file`,
  )
  .option('--schema <path>', `[string] prisma schema path`, {
    default: './prisma/schema.prisma',
  })
  .action(createApis)

cli
  .command('dapi [route] [alias]', 'Delete API')
  .option('-t,--template-alias <string>', `[string] template alias`)
  .option('--with-test', '[boolean] Generate test file too')
  .action(deleteApis)

cli
  .command('gdoc', 'Generate API document (json format)')
  .option('-o,--output-dir <path>', `[string] output dir`, {
    default: 'docs',
  })
  .option('--tsconfig <path>', `[string] tsconfig file path`, {
    default: 'tsconfig.json',
  })
  .action(startGenerateDoc)

cli
  .command(
    'sapi',
    'Sync API, update all relative files according to files in --api-dir',
  )
  .action(syncApiByFile)

cli
  .command('gpure', 'Generate purifed mutation types from schema.prisma')
  .option('-i,--input <path>', `[string] input path`, {
    default: './prisma/schema.prisma',
  })
  .option('-o,--output <path>', `[string] output path`, {
    default: './types/pure.d.ts',
  })
  .option('-n,--namespace <name>', `[string] Namespace of purified types`, {
    default: 'Pure',
  })
  .action(generatePureTypes)

// Listen to unknown commands
cli.on('command:*', () => {
  console.error('Invalid command: %s', cli.args.join(' '))
  process.exit(1)
})

cli.help()

export { cli }
