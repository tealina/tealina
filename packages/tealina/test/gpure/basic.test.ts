import { expect, test } from 'vitest'
import { workflow } from '../../src/commands/gpure'

const snapshot = [
  '/**  public comment */',
  'interface UserCreateInput{',
  '  /** @default {autoincrement()} */',
  '  id?: number',
  '  /** @default {now()} */',
  '  createdAt?: number',
  '  /**  inline public comment */',
  '  email?: string',
  '  name?: string',
  '}',
  '',
  '/**  public comment */',
  'interface UserUpdateInput{',
  '  /** @default {autoincrement()} */',
  '  id?: number',
  '  /** @default {now()} */',
  '  createdAt?: number',
  '  /**  inline public comment */',
  '  email?: string',
  '  name?: string',
  '  /** @default {USER} */',
  '  role?: any',
  '}',
  '',
  'interface PostCreateInput{',
  '  /** @default {autoincrement()} */',
  '  id?: number',
  '  /** @default {now()} */',
  '  createdAt?: number',
  '  updatedAt?: number',
  '  /** @default {false} */',
  '  published?: boolean',
  '  title: string',
  '  authorId?: number',
  '}',
  '',
  'interface PostUpdateInput{',
  '  /** @default {autoincrement()} */',
  '  id?: number',
  '  /** @default {now()} */',
  '  createdAt?: number',
  '  updatedAt?: number',
  '  /** @default {false} */',
  '  published?: boolean',
  '  title?: string',
  '  authorId?: number',
  '}',
  '',
  '/**',
  '* Enums',
  '*/',
  '// Based on',
  '// https://github.com/microsoft/TypeScript/issues/3192#issuecomment-261720275',
  '',
  'export const Role: {',
  '  USER: "USER"',
  '  ADMIN: "admin"',
  '}',
  'export type Role = (typeof Role)[keyof typeof Role]',
  '',
]

test('relation database', async () => {
  const result = await workflow('test/utils/mock/mock.prisma', {
    typeRemap: t => (t == 'DateTime' ? 'number' : null),
    overwrite: {
      excludeProps: [
        {
          blockName: 'User',
          keyword: 'model',
          kind: 'CreateInput',
          predicate: p => p.name == 'role',
        },
      ],
      isOptional: [
        {
          blockName: 'User',
          keyword: 'model',
          kind: 'CreateInput',
          predicate: p => p.name == 'email',
        },
      ],
    },
  })
  expect(result).deep.eq(snapshot)
})
