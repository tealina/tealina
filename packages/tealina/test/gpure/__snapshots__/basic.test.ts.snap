// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`relation database 1`] = `
[
  "/**  public comment */",
  "interface User {",
  "  /** @default {autoincrement()} */",
  "  id: number",
  "  /** @default {now()} */",
  "  createdAt: Date",
  "  /**  inline public comment */",
  "  email: string",
  "  name: string | null",
  "  /** @default {USER} */",
  "  role: Role",
  "}",
  "/**  public comment */",
  "interface UserCreateInput {",
  "  /** @default {now()} */",
  "  createdAt?: number",
  "  /**  inline public comment */",
  "  email?: string",
  "  name?: string",
  "  /** @default {USER} */",
  "  role?: Role",
  "}",
  "/**  public comment */",
  "interface UserUpdateInput {",
  "  /** @default {autoincrement()} */",
  "  id?: number",
  "  /** @default {now()} */",
  "  createdAt?: number",
  "  /**  inline public comment */",
  "  email?: string",
  "  name?: string",
  "  /** @default {USER} */",
  "  role?: Role",
  "}",
  "interface Post {",
  "  /** @default {autoincrement()} */",
  "  id: number",
  "  /** @default {now()} */",
  "  createdAt: Date",
  "  updatedAt: Date",
  "  /** @default {false} */",
  "  published: boolean",
  "  title: string",
  "  authorId: number | null",
  "}",
  "interface PostCreateInput {",
  "  /** @default {now()} */",
  "  createdAt?: number",
  "  updatedAt?: number",
  "  /** @default {false} */",
  "  published?: boolean",
  "  title: string",
  "  authorId?: number",
  "}",
  "interface PostUpdateInput {",
  "  /** @default {autoincrement()} */",
  "  id?: number",
  "  /** @default {now()} */",
  "  createdAt?: number",
  "  updatedAt?: number",
  "  /** @default {false} */",
  "  published?: boolean",
  "  title?: string",
  "  authorId?: number",
  "}",
  "/**",
  "* Enums",
  "*/",
  "// Based on",
  "// https://github.com/microsoft/TypeScript/issues/3192#issuecomment-261720275",
  "",
  "export const Role: {",
  "  USER: \\"USER\\"",
  "  ADMIN: \\"admin\\"",
  "}",
  "export type Role = (typeof Role)[keyof typeof Role]",
  "",
]
`;
